### ISSUE TITLE
Add support for 2FA

### ISSUE THREAD

**Issue by rbnis** [inwx/ansible-collection#3]
Hello,

I currently use 2FA for my account and would prefer to not disable it.
Unfortunatly 2FA is currently unsupported, as stated here: https://github.com/inwx/ansible-collection/blob/master/plugins/modules/dns.py#L27

Are there any plans to support this in the near future?

Also, I'm happy to help out with this if you like.

**Comment by NickUfer** [comment]
Hey, as stated in the notes, the API client is rebuilt every time a task is performed. If we supported 2FA, each task would have a delay of approximately 30 seconds. This is because the RFC prohibits the use of a token twice and a new token is only available every 30 seconds: https://tools.ietf.org/html/rfc6238#section-5.2
```
Note that a prover may send the same OTP inside a given time-step
window multiple times to a verifier. The verifier MUST NOT accept
the second attempt of the OTP after the successful validation has
been issued for the first OTP, which ensures one-time only use of an
OTP.
```

One solution we considered would be to save the session to a file and reuse it between tasks. However, we still have to evaluate this solution further. I will inform you once we have found a solution :+1:

**Comment by Robinhuett** [comment]
I guess this makes sense.
One downside to the session file would be that it would need to be cached by the CD system. Not hard to do, but a user has to do it.

As far as I can tell, there are also no API tokens one could use at INWX. I don't know how much influence you have on the backend side here, but I think this would be an good solution. Even though it's also the most complex 🙈 

Thank you for investigating this :)

**Comment by DC7IA** [comment]
I agree. It'd be great to have API tokens, preferrably with limited access (i.e. to specific domain and record types) for access to the API.

**Comment by markuman** [comment]
Yes, that's a must-have feature.  

API tokens, or a dedicated module that creates a session that is only saved via `register: inwx_session` that is passed than to the `inwx.collection.dns` module as `session: inwx_session` parameter.

**Comment by XenGi** [comment]
Any news on this? Seems like there is still nothing like that in the module. But there is a `shared_secret` param in the login function so I guess preparation work has begun?

**Comment by NickUfer** [comment]
The feature was already implemented but I forgot that it is not compatible with the way ansible works. Every task creates a new instance of this module. You theoretically must login again with every task with the credentials and the OTP. But a single OTP can only be used once. After that you have to wait up to 30 seconds for the next OTP you can use.
The problem with logging in for every request without an OTP is bypassed by sending the login credentials with every API request. If you send them every time you don't have to call account.login and the authentication happens with every request transparently.

The solutions for this could be:
* We finally implement API tokens (Planned but still far away because of the existing code base)
* The solution of @markuman (sounds good to me)

I must admit I haven't looked into for a long time as I hit multiple road blocks. I hope I can finally find time to fix all the open issues with this project in the coming 3 weeks with input from our sysadmins who use this module.

**Comment by leegarrett** [comment]
IMHO this is the wrong approach. ansible has very little support for interaction (only I can think of is -K, "ask for privilege escalation password", which might already be used for the playbook), and even that is problematic as TOTP is time-critical. Instead there should be the option to create API keys in the web frontend and use those. Bonus points if you can allow/deny permissions for every API key.

That is current industry practice, so I'd go with that.

**Comment by markuman** [comment]
You can enter the TOTP via `vars_prompt:` at the beginning of the play and the first task creates the session. That's something thousands of users are doing it multiple times in the AWS world with STS tokens.   
And in reality, TOTP is not that strict. there are often 2-5 30s windows configured where the TOTP is valid.   
Another possibility is, that the TOTP comes from another lookup provider, so there is no need to save it at the same place as the password is stored. Or vice-versa. The password is stored in some password manager (_1password, nextcloud passwords, hashicorp vault, you name it_) while the TOTP secret is kept in an ansible vault.

**Comment by NickUfer** [comment]
I have added an `inwx.collection.session` module as suggested by @markuman . It is available in the `1.3.0` release. See the following document for usage guidance: https://github.com/inwx/ansible-collection/blob/1.3.0/playbooks/examples/dns_use_2fa.yml

And of course as already said in this thread: don't save username + password and shared secret in the same place as this cancels all benefits of having two factors for authentication.
